[WPE] WPE Platform: add WPEDRMDevice

carlosgcampos opened the pull request:
#### a2b0df6f37af5235490c39a2615dcf6d7a6b67e7
<pre>
[WPE] WPE Platform: add WPEDRMDevice
<a href="https://bugs.webkit.org/show_bug.cgi?id=296682">https://bugs.webkit.org/show_bug.cgi?id=296682</a>

Reviewed by Adrian Perez de Castro.

Instead of exposing DRM nodes individually, add WPEDRMDevice with
getters for primary and render node. Make wpe_display_get_drm_device()
return a WPEDRMDEvice instead of the path of a node and remove
wpe_display_get_drm_render_node(). Also change WPEBufferDMABufFormats to
use WPEDRMDEvice for the main and target devices instead of paths for
nodes.

Canonical link: <a href="https://commits.webkit.org/298125@main">https://commits.webkit.org/298125@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e4de7f476afc2f3f90c03a560ad85d3d7dba949

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114341 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24550 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120505 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65059 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34717 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42647 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86890 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117289 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27649 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102690 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67282 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26830 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20815 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64195 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97024 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20931 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123715 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41355 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30841 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95714 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41732 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98890 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95498 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40650 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18487 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37397 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18321 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41235 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46743 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40830 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44136 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42578 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEBufferDMABufFormats.cpp:
```diff
@@ -53,24 +53,29 @@ struct DMABufFormat {
     }
 
-    CString device;
+    WPEDRMDevice* device { nullptr };
```
Could we add a `GRefPtrWPEPlatform.{h,cpp}` with the template specialization to use `GRefPtr<WPEDRMDevice>`? Then it could be used here, simplifying the move constructor, and the destructor could be left untouched (and let the compiler generate it).

carlosgcampos replied:
Yes, for WPEDRMDevice and others I guess, so maybe it would be better in a follow up.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEBufferDMABufFormats.cpp:
```diff
@@ -232,12 +247,17 @@ GArray* wpe_buffer_dma_buf_formats_get_format_modifiers(WPEBufferDMABufFormats*
     }
 
-    CString device;
+    WPEDRMDevice* device { nullptr };
```
Here we could also use said `GRefPtr<WPEDRMDevice>`.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
```diff
@@ -66,6 +67,7 @@ struct _WPEDisplayPrivate {
     GRefPtr<WPEClipboard> clipboard;
     GRefPtr<WPESettings> settings;
     WPEAvailableInputDevices availableInputDevices;
+    WPEDRMDevice* overridenDRMDevice;
```
...and this one could be a smart pointer, too.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEDRMDevicePrivate.h:
```diff
@@ -22,12 +22,9 @@
+#include "WPEDRMDevice.h"
 
-#endif /* USE(LIBDRM) */
+WPEDRMDevice* wpeDRMDeviceCreateForDevice(const char*);
```
This helper function could return a `GRefPtr<WPEDRMDevice>`.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -51,8 +53,8 @@
-    CString drmDevice;
-    CString drmRenderNode;
+    WPEDRMDevice* displayDevice;
+    WPEDRMDevice* renderDevice;
```
...and here, too.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp:
```diff
@@ -55,23 +55,26 @@
  *
  */
 struct _WPEDisplayHeadlessPrivate {
+    std::optional<WPEDRMDevice*> drmDevice;
```
It's redundant to use `std::optional` for a pointer, because we can check whether the pointer is null instead.

carlosgcampos replied:
Not in this case, because we don't know if nullptr means we didn't initialize it or we initialized it but we got nullptr.  So, nullopt means not initialized, nullptr means it's initialized to nullptr.

aperezdc reviewed (commented):
Nice, I like that this makes the code clearer overall, and brings in a few simplifications.

OTOH, it looks to me like it would be worth it to provide a `GRefPtr<WPEDRMDevice>` template specialization, which would further refine and help simplify the code.

aperezdc reviewed (approved):
Thanks, this is nicer now with the smart pointers 👍🏼

carlosgcampos committed:
[WPE] WPE Platform: add WPEDRMDevice
https://bugs.webkit.org/show_bug.cgi?id=296682

Reviewed by Adrian Perez de Castro.

Instead of exposing DRM nodes individually, add WPEDRMDevice with
getters for primary and render node. Make wpe_display_get_drm_device()
return a WPEDRMDEvice instead of the path of a node and remove
wpe_display_get_drm_render_node(). Also change WPEBufferDMABufFormats to
use WPEDRMDEvice for the main and target devices instead of paths for
nodes.

Canonical link: https://commits.webkit.org/298125@main
